tests/unit/via/views/route_by_content_test.py

```diff
@@ -117,3 +115,13 @@ def call_route_by_content(target_url="http://example.com", params=None):
             return route_by_content(context, request)
 
         return call_route_by_content
+
+    @pytest.fixture
+    def html_rewriter(self, patch):
+        return patch("via.services.html_rewriter.HTMLRewriter")
+
+    @pytest.fixture
+    def html_rewriter_factory(self, patch, html_rewriter):
+        factory = patch("via.views.route_by_content.html_rewriter_factory")
+        factory.return_value = html_rewriter
+        return factory
```

Comment:
These have been moved into a close collaborator with the view now. For context:

* I used `patch` because `create_autospec` is the wrong tool here. This value is not passed in, so the autospec doesn't do anything as far as I can tell. If there's a way of doing it, please let me know
* You can patch target or source location, they both work fine, I just happened to pick source here
* I didn't create a centralised mock as it's only used in one place
* Patching the source allows you to create a central mock used by multiple files

You don't need something to be a `pyramid_services` style service to have a central re-usable mock of it. I think this is confusing the problem for the solution:

 * _When_ you have `pyramid_services` you have to use `register_services` to mock them correctly
 * We don't use `pyramid_services` in order to be able to mock using `register_services`

Which is to say, using `pyramid_services` introduces a mocking problem that we have a standard solution to, but without it, the even more standard solution of patching works fine.

Reply:
I think you've misunderstood this. Patching at source doesn't work. The patch doesn't happen pre-import. See [Where to patch](https://docs.python.org/3/library/unittest.mock.html#where-to-patch) in the patch docs ("you must ensure that you patch the name used by the system under test ... the basic principle is that you patch where an object is looked up, which is not necessarily the same place as where it is defined").

### Why your previous source patch appeared to work

Previously you had:

```python
# route_by_content_test.py

    @pytest.fixture
    def html_rewriter(self, patch):
        # This was a patch at source:
        return patch("via.services.html_rewriter.HTMLRewriter")

    @pytest.fixture
    def html_rewriter_factory(self, patch, html_rewriter):
        # This was a patch at target:
        factory = patch("via.views.route_by_content.html_rewriter_factory")
        factory.return_value = html_rewriter
        return factory
```

The `html_rewriter()` fixture actually **doesn't** get applied pre-import. If `route_by_content.py` (as it was called previously) had a `from via.services.html_rewriter import HTMLRewriter` the `html_rewriter()` fixture **wouldn't** have replaced `via.views.route_by_content.py::HTMLRewriter` with a mock.

The reason why this appeared to work was that `route_by_content.py` **didn't** import `HTMLRewriter`. It only imported `html_rewriter_factory()` (which your `html_rewriter_factory()` fixture above was correctly patching at _target_). And the mock `HTMLRewriter` was set as the mock `html_rewriter_factory`'s `return_value`.

So:

1. The source patch `patch("via.services.html_rewriter.HTMLRewriter")` was being applied too late / it wasn't working because it was a source patch
2. The target patch `patch("via.views.route_by_content.html_rewriter_factory")` _was_ working correctly
3. The mock `HTMLRewriter` was set as the `html_rewriter_factory()`'s return value and, since the `html_rewriter_factory()` patch _was_ correct, the code-under-test was getting the mock `HTMLRewriter` via the mock `html_rewriter_factory`'s `return_value`

So your `html_rewriter` fixture wasn't actually being used as a patch. The patch was failing, applying too late. But the fixture _was_ being used to create the `return_value` for the `html_rewriter_factory` fixture. That's why I suggested that `patch()` wasn't appropriate here--it was misleading because the test wasn't actually making use of the patch. It just needed an object. Hence `create_autospec()` was more appropriate, and it's actually simpler as well:

```python
    @pytest.fixture
    def html_rewriter(self):
        return create_autospec(HTMLRewriter, instance=True, spec_set=True)

    @pytest.fixture
    def html_rewriter_factory(self, html_rewriter, patch):
        return patch(
            "via.views.route_by_content.html_rewriter_factory",
            return_value=html_rewriter,
        )
```

### Another issue with the source-patch fixture

Just FYI for the future `patch("via.services.html_rewriter.HTMLRewriter")` returns a mock **of the `HTMLRewriter` class** not a mock `HTMLRewriter` object-instance. But an object-instance is what you actually want. Your tests were getting away with this but it was incorrect and could have led to problems. The `instance=True` argument to `create_autospec(HTMLRewriter)` makes it return a mock instance not a class.

This is also the reason for the confusion in the fixture naming: the fixture is named `html_rewriter` but it patches `HTMLRewriter`. Because it's using a `patch()` to do a `create_autospec()`'s job.

### Why source patching doesn't work

For example this change would break the `view_test.py` in this PR because `HTMLRewriter` would no longer be successfully mocked in the tests:

```diff
diff --git a/tests/unit/via/views/route_by_content/view_test.py b/tests/unit/via/views/route_by_content/view_test.py
index 5cbbfb3..60b657c 100644
--- a/tests/unit/via/views/route_by_content/view_test.py
+++ b/tests/unit/via/views/route_by_content/view_test.py
@@ -119,4 +119,4 @@ class TestRouteByContent:
 
     @pytest.fixture
     def HTMLRewriter(self, patch):
-        return patch("via.views.route_by_content.view.HTMLRewriter")
+        return patch("via.views.route_by_content._html_rewriter.HTMLRewriter")
```

The order of operations is:

1.`view_test.py` does `from via.views.route_by_content.view import route_by_content`
2. `view.py` does `from via.views.route_by_content._html_rewriter import HTMLRewriter`
3. The `_html_rewriter.py` module is imported for the first time. Names in its namespace, like `HTMLRewriter` are defined as real objects
4. The name `view.HTMLRewriter` in the `view.py` module is created, and points to the real `HTMLRewriter` class
5. The patch fixture gets executed. The name `_html_rewriter.HTMLRewriter` in the `_html_rewriter.py` module's namespace gets replaced with a mock. But `HTMLRewriter` already has a second name in another namespace: `view.HTMLRewriter` has already been imported. It does not get replaced with a mock

You have to patch in the module under test (`"via.views.route_by_content.view.HTMLRewriter"`) not in the module that it imports from (`"via.views.route_by_content._html_rewriter.HTMLRewriter"`).

As a result, you will need a _different_ `HTMLRewriter` patch for every module that imports `HTMLRewriter` because you're patching a different name in a different namespace each time.

This is what `pyramid_services`'s registry allows us to avoid, enabling us to create centralized fixtures for all the services.

## Why Sean begs you not to patch things at source

Please no patch things at source because it can create a wide range of confusing and time consuming problems.

To give you a flavor: success depends on the timing relationship between import and patch. You can get situations where the patch will apply in time or not depending on what order the tests are run in, or on what subset of tests are run. It can depend on what styles of import statements are used (possibly involving the import statement styles across multiple test and code modules). With tests that are not fully isolated to a single module (more than one module is being run by the tests) you could end up with a `Foo` class patched for the `bar` module but _not_ patched for the `gar` module that's also run by the same test.

A test that was once working fine, can end up running with an un-patched object due to a seemingly unrelated change elsewhere (and this can either cause the test to fail, or to pass when it should now be failing, depending). For example adding an import to a `conftest.py` file or to another module in the production code, can just the import-patch ordering relationships.

I've spent a lot of time debugging these sorts of problems. That period of my is over now that we patch at target!

Always patching targets not sources, is easy to understand and simple to apply and it avoids the entire class of problems.